### DISCUSSION TITLE
`pg_class_row` with JOIN

### DISCUSSION THREAD

**Discussion by matheusssilva** [psycopg/psycopg#908] [Q&A] [answered]
I would like to know if the row factory functionality supports the creation of composite objects using the JOIN clause. For example.

```
class Profile:

    id: int | None = None
    nome: str | None = None
    permissions: list[bool] | None = None


class User:

    id: int | None = None
    name: str | None = None
    profile_id: int | None = None
    profile: Profile | None = None
```

What would the **SQL** code have to be like for `fatchone()` to return a `User` object like:

```
User(
    1,
    "Any User Name",
    1,
    Profile(
        1,
        "Any Profile Name",
        [false, false, true]
    )
)
```

I am passing the User class to the `pg_class_row(User)` function.

**Comment by dlax** [comment, marked as answer]
Assuming you're executing the following query:
~~~sql
SELECT
  u.id AS user_id, u.name AS user_name, p.name AS profile_name, p.permissions AS profile_permissions
FROM
  users u JOIN profiles p ON u.profile_id = p.id
~~~
you can then define a "factory" function as follows (could also be a class method on `User` class):
~~~python
def user_from_row(*, user_id, user_name, profile_name, profile_permissions):
    return User(user_id, user_name, Profile(profile_name, profile_permissions))
~~~
where `user_from_row` would be argument for [psycopg.rows.kwargs_row](https://www.psycopg.org/psycopg3/docs/api/rows.html#psycopg.rows.kwargs_row) to be used as:
~~~python
>>> cur = conn.cursor(row_factory=kwargs_row(user_from_row))
>>> cur.execute(<the query above>)
>>> cur.fetchone()
# >  returns the expected User(...) object
~~~

**Reply by matheusssilva** [reply]
So, the `pg_class_row` function not able to generate factories for composite classes? Only simple entities without relationship?

**Reply by dlax** [reply]
If you define the `__init__()` method of your class similarly to the `user_from_row()` function suggested above, it should work:
~~~python
class User:
  ...
  def __init__(self,*, user_id, user_name, profile_name, profile_permissions):
    self.id = user_id
    self.name = user_name
    self.profile = Profile(profile_name, profile_permissions)
~~~
But, no, there's "automatic" way to handle your use case.

**Reply by matheusssilva** [reply]
In this case, I believe the best way is to create a separate factory function. Changing the way domains are initialized would impact the entire business rule.
